Match FontFaceSet font loading with spec

AlexsanderDamaceno opened the pull request:
#### 04a74bd9c97a07e7428c2df4794ad0f3532099a8
<pre>
Match FontFaceSet font loading with spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=280954">https://bugs.webkit.org/show_bug.cgi?id=280954</a>

Reviewed by Tim Nguyen.

As per spec <a href="https://drafts.csswg.org/css-font-loading/#switch-the-fontfaceset-to-loading">https://drafts.csswg.org/css-font-loading/#switch-the-fontfaceset-to-loading</a>,
steps 3 and 4 were missing in the FontFaceSet font-loading algorithm, which caused, for example,
loading two diferrent fonts using promises to not work and not fire loading event.

* Source/WebCore/css/FontFaceSet.cpp:
(WebCore::FontFaceSet::startedLoading):
* Source/WebCore/css/FontFaceSet.h:
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-loadingevent-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-loadingevent.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-fonts-loading-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-fonts-loading.html: Added.

Canonical link: <a href="https://commits.webkit.org/301471@main">https://commits.webkit.org/301471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/078e087ebfe88f404e5859a7dbb7c899bf1b7661

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126060 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/45713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132883 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77874 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127931 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46383 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54258 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96002 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64101 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129008 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37108 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/112735 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76491 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36013 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30920 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76355 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106890 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31140 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135583 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52817 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40547 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104495 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53274 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108953 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104213 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49613 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27942 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50195 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19718 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/52714 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58547 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52048 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55395 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/53752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cc86436f63bcdbc91d045e09eedc11d39b07499d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/34754

nt1m commented on Source/WebCore/css/FontFaceSet.cpp:
```diff
@@ -261,7 +261,7 @@ void FontFaceSet::faceFinished(CSSFontFace& face, CSSFontFace::Status newStatus)
 
 void FontFaceSet::startedLoading()
 {
-    // FIXME: Fire a "loading" event asynchronously.
```
Not sure how much it matters, but your code fires the event synchronously (which may be correct depending on what the comment is relative to)

nt1m reviewed (commented):
This needs a test (preferably web-platform-test)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f55c5bb50ca94fea0c75380e3c2a87ae2bb4d227. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/34754

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c7e8c70bcd36628fa2fd257acc00467b34174092. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/34754

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/37f0914d3d120aad286eda46427f5d2d6993fb33. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/34754

vitorroriz reviewed (commented):
Can you also test it via 'onloading', please?

nt1m commented:
We should probably also fix https://bugs.webkit.org/show_bug.cgi?id=299849 before emitting the event

AlexsanderDamaceno committed:
Match FontFaceSet font loading with spec
https://bugs.webkit.org/show_bug.cgi?id=280954

Reviewed by Tim Nguyen.

As per spec https://drafts.csswg.org/css-font-loading/#switch-the-fontfaceset-to-loading,
steps 3 and 4 were missing in the FontFaceSet font-loading algorithm, which caused, for example,
loading two diferrent fonts using promises to not work and not fire loading event.

* Source/WebCore/css/FontFaceSet.cpp:
(WebCore::FontFaceSet::startedLoading):
* Source/WebCore/css/FontFaceSet.h:
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-loadingevent-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-loadingevent.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-fonts-loading-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-font-loading/fontface-fonts-loading.html: Added.

Canonical link: https://commits.webkit.org/301471@main
